File: src/py/models/dft.py
"""DFT model implementation."""

from typing import Any, Dict, List, Optional
from ..model import Model


class DFTModel(Model):
    """DFT model class extending base Model."""
    
    def __init__(self, config: Dict[str, Any]):
        """Initialize DFTModel.
        
        Args:
            config: Model configuration
        """
        super().__init__(**config)
        
        # Import here to avoid circular imports
        method_factory = config.get('MethodFactory')
        if method_factory:
            from ..methods.factory import MethodFactory
            self._method_factory = method_factory
        else:
            from ..methods.factory import MethodFactory
            self._method_factory = MethodFactory
    
    @property
    def group_slug(self) -> str:
        """Build slug string based on model information.
        
        Returns:
            str: Group slug
        """
        parts = [
            self._application['shortName'],
            self.type,
            self.subtype,
            self.functional['slug'],
            '+'.join(r['slug'] for r in self.refiners),
            '+'.join(m['slug'] for m in self.modifiers),
        ]
        
        # Join with ':' and clean up empty parts
        slug = ':'.join(part for part in parts if part)
        slug = slug.replace('::', ':').rstrip(':')
        return slug
    
    @property
    def default_functional(self) -> Dict[str, str]:
        """Get default functional.
        
        Returns:
            Dict: Default functional with slug and name
        """
        from ..tree import tree_slug_to_named_object
        functionals = self.tree_branch_for_subtype.get('functionals', [])
        if functionals:
            return tree_slug_to_named_object(functionals[0])
        return {'slug': '', 'name': ''}
    
    @property
    def default_refiners(self) -> List[Dict[str, str]]:
        """Get default refiners.
        
        Returns:
            List: Default refiners list
        """
        return []
    
    @property
    def default_modifiers(self) -> List[Dict[str, str]]:
        """Get default modifiers.
        
        Returns:
            List: Default modifiers list
        """
        return []
    
    @property
    def functional(self) -> Dict[str, str]:
        """Get functional.
        
        Returns:
            Dict: Functional configuration
        """
        return self.prop('functional', self.default_functional)
    
    @property
    def refiners(self) -> List[Dict[str, str]]:
        """Get refiners.
        
        Returns:
            List: Refiners list
        """
        return self.prop('refiners', self.default_refiners)
    
    @property
    def modifiers(self) -> List[Dict[str, str]]:
        """Get modifiers.
        
        Returns:
            List: Modifiers list
        """
        return self.prop('modifiers', self.default_modifiers)
    
    def set_subtype(self, subtype: str) -> None:
        """Set model subtype.
        
        Args:
            subtype: New subtype value
        """
        self.set_prop('subtype', subtype)
        self.set_functional(self.default_functional)
    
    def set_functional(self, functional: Dict[str, str]) -> None:
        """Set functional.
        
        Args:
            functional: Functional configuration
        """
        self.set_prop('functional', self._string_to_slugified_object(functional))
        self.set_method(self._method_factory.create(self.default_method_config))
    
    def _set_array_prop(self, name: str, data: List[Any]) -> None:
        """Set array property.
        
        Args:
            name: Property name
            data: Data to set
        """
        if not isinstance(data, list):
            data = [data]
        
        slugified_data = [self._string_to_slugified_object(item) for item in data]
        self.set_prop(name, slugified_data)
        setattr(self, f'_{name}', slugified_data)
    
    def set_refiners(self, refiners: List[Dict[str, str]]) -> None:
        """Set refiners.
        
        Args:
            refiners: Refiners list
        """
        self._set_array_prop('refiners', refiners)
    
    def set_modifiers(self, modifiers: List[Dict[str, str]]) -> None:
        """Set modifiers.
        
        Args:
            modifiers: Modifiers list
        """
        self._set_array_prop('modifiers', modifiers)
    
    def to_json(self) -> Dict[str, Any]:
        """Convert to JSON representation.
        
        Returns:
            Dict: JSON representation
        """
        base_json = super().to_json()
        
        # Only store slug for functional
        functional_slug = {'slug': self.functional['slug']} if self.functional else {}
        
        return {
            **base_json,
            'functional': functional_slug,
            'refiners': self.refiners,
            'modifiers': self.modifiers,
        }
    
    @property
    def all_functionals(self) -> List[Dict[str, str]]:
        """Get all functionals in the form of {name: ..., slug: ...}.
        
        Returns:
            List: List of functional objects
        """
        from ..tree import tree_slug_to_named_object
        functionals = self.tree_branch_for_subtype.get('functionals', [])
        return [tree_slug_to_named_object(x) for x in functionals]
    
    @property
    def all_refiners(self) -> List[Dict[str, str]]:
        """Get all refiners in the form of {name: ..., slug: ...}.
        
        Returns:
            List: List of refiner objects
        """
        from ..tree import tree_slug_to_named_object
        refiners = self.tree_branch_for_subtype.get('refiners', [])
        return [tree_slug_to_named_object(x) for x in refiners]
    
    @property
    def all_modifiers(self) -> List[Dict[str, str]]:
        """Get all modifiers in the form of {name: ..., slug: ...}.
        
        Returns:
            List: List of modifier objects
        """
        from ..tree import tree_slug_to_named_object
        modifiers = self.tree_branch_for_subtype.get('modifiers', [])
        return [tree_slug_to_named_object(x) for x in modifiers]

File: src/js/models/dft.ts
import { safeMakeArray } from "@mat3ra/code/dist/js/utils";
import type { AnyObject } from "@mat3ra/esse/dist/js/esse/types";
import {
    type DFTModelSchema,
    SlugifiedEntry,
    SlugifiedEntryOrSlug,
} from "@mat3ra/esse/dist/js/types";

import { MethodFactory } from "../methods/factory";
import { Model } from "../model";
import { treeSlugToNamedObject } from "../tree";
import type { ModelConfig } from "../types";

export class DFTModel extends Model {
    constructor(config: ModelConfig & { MethodFactory?: typeof MethodFactory }) {
        super(config);
        this._MethodFactory = config.MethodFactory || MethodFactory;
    }

    declare type: DFTModelSchema["type"];

    declare subtype: DFTModelSchema["subtype"];

    get groupSlug(): string {
        const refinersSlug = this.refiners.map((o) => o.slug).join("+");
        const modifiersSlug = this.modifiers.map((o) => o.slug).join("+");
        const slugs = [
            this._application?.shortName,
            this.type,
            this.subtype,
            this.functional,
            refinersSlug,
            modifiersSlug,
        ].filter(Boolean);
        return slugs.join(":");
    }

    get defaultFunctional(): SlugifiedEntry {
        const [slug] = this.treeBranchForSubType.functionals || [];
        return treeSlugToNamedObject(slug);
    }

    readonly defaultRefiners: SlugifiedEntry[] = [];

    readonly defaultModifiers: SlugifiedEntry[] = [];

    get slugifiedFunctional() {
        return this._stringToSlugifiedObject(this.functional);
    }

    get functional() {
        return this.requiredProp<DFTModelSchema["functional"]>("functional");
    }

    get refiners() {
        return this.prop<SlugifiedEntry[]>("refiners", this.defaultRefiners);
    }

    get modifiers() {
        return this.prop<SlugifiedEntry[]>("modifiers", this.defaultModifiers);
    }

    setSubtype(subtype: SlugifiedEntryOrSlug): void {
        this.setProp("subtype", subtype);
        this.setFunctional(this.defaultFunctional);
    }

    setFunctional(functional: SlugifiedEntryOrSlug): void {
        this.setProp("functional", this._stringToSlugifiedObject(functional));
        this.setMethod(this._MethodFactory.create(this.defaultMethodConfig));
    }

    private _setArrayProp(
        name: "refiners" | "modifiers",
        data: SlugifiedEntryOrSlug | SlugifiedEntryOrSlug[],
    ): void {
        const normalized = safeMakeArray(data).map((item) => this._stringToSlugifiedObject(item));
        this.setProp(name, normalized);
        (this as unknown as Record<string, SlugifiedEntry[]>)[`_${name}`] = normalized;
    }

    setRefiners(refiners: SlugifiedEntryOrSlug | SlugifiedEntryOrSlug[]): void {
        this._setArrayProp("refiners", refiners);
    }

    setModifiers(modifiers: SlugifiedEntryOrSlug | SlugifiedEntryOrSlug[]): void {
        this._setArrayProp("modifiers", modifiers);
    }

    toJSON(): DFTModelSchema & AnyObject {
        const baseJson = super.toJSON();
        const keysToExclude = ["type", "subtype", "functional", "refiners", "modifiers", "method"];
        const restJson = Object.fromEntries(
            Object.entries(baseJson).filter(([key]) => !keysToExclude.includes(key)),
        );

        return {
            type: this.type,
            subtype: this.subtype,
            method: this.Method.toJSONWithCleanData(),
            functional: this.functional,
            refiners: this.refiners,
            modifiers: this.modifiers,
            ...restJson,
        } as DFTModelSchema;
    }

    get allFunctionals(): SlugifiedEntry[] {
        return (this.treeBranchForSubType.functionals || []).map((slug) =>
            treeSlugToNamedObject(slug),
        );
    }

    get allRefiners(): SlugifiedEntry[] {
        return (this.treeBranchForSubType.refiners || []).map((slug) =>
            treeSlugToNamedObject(slug),
        );
    }

    get allModifiers(): SlugifiedEntry[] {
        return (this.treeBranchForSubType.modifiers || []).map((slug) =>
            treeSlugToNamedObject(slug),
        );
    }
}
